### ISSUE TITLE
Don't allocate pty if docker command output is redirected to regular file

### ISSUE THREAD

**Issue by praiskup** [docker/docker#17347]
Example:
```
$ docker run --rm -i test cat /usr/bin/jej
#! /bin/sh

debug()
{
    echo "$*"
}

for i in 0 1 2
do
  test -t $i && debug "$i is a tty" || debug "$i is not a tty"
done
$ docker run --rm -i test /usr/bin/jej
0 is not a tty
1 is not a tty
2 is not a tty
$ docker run --rm -it test /usr/bin/jej 
0 is a tty
1 is a tty
2 is a tty
$ docker run --rm -it test /usr/bin/jej 2>/dev/null
0 is a tty
1 is a tty
2 is a tty
$ docker run --rm -it test /usr/bin/jej >/dev/null
$ docker run --rm -it test /usr/bin/jej > stdout
$ cat stdout
0 is a tty
1 is a tty
2 is a tty
```

This would be useful for containerized command output colorization -- if the output is
redirected to non-tty file -> avoid shell colors in output.

**Comment by LK4D4** [comment]
I dunno why to use `-it` with non-interactive usage. Issue looks like inactive anyway.

**Comment by praiskup** [comment]
This makes the docker output behave wrongly, and makes writing script really inconvenient...

By default, unix terminal is interactive, and most of the terminal utilities expect that.  Actuall, the '-t' option is a bit overkill, because you can detect that automatically within docker command.

Due to this bug, command-line outputs can not be colorized the "good way":
```
$ alias useful_command='docker --run -ti XXX command'
$ useful_command
colorized output
$ useful_command > stdout
$ cat stdout
not-colorized, not broken output
```

**Comment by praiskup** [comment]
The unfortunate thing about github is that, as an original reporter, I'm unable to re-open the issue.
So now we probably forget about this issue?

**Comment by LK4D4** [comment]
@praiskup I encourage you to open new issue and describe real use cases there. 
Nor title, neither discussion match your report IMO - calling tty command is not real use case, that command has bug about working in containers and sort of "synthetic". 
I'd prefer some real script with coloring and demonstration how it behave badly with docker.
Thanks!

**Comment by praiskup** [comment]
>  IMO - calling tty command is not real use case

I would like to think that this is not truth.

> I'd prefer some real script with coloring and demonstration how it behave badly with docker.

Please don't overestimate this, this is just small bug I requested fix for.  But thanks for this comment, I'll hack the script.

**Comment by praiskup** [comment]
For example, `GNU tar` by default put's archive into standard output.  Unfortunately, our `Fedora` image is trimed enough to not have GNU tar in default installation, so you need to have Dockerfile, or run `ubuntu` image:
```
$ docker run -ti --rm ubuntu tar c /etc/
tar: Refusing to write archive contents to terminal (missing -f option?)
tar: Error is not recoverable: exiting now
```
That is definitely expected, `tar` just warns us that putting binary content out into terminal could cause that the terminal is no longer useful...  So what that means is that tar is asking for file descriptor redirection, usually `tar c /etc/ > /tmp/archive`.  But with docker:

```
$ docker run -ti --rm ubuntu tar c /etc/ > /tmp/archive
[[ no error reported ]]
```
That actually looks like tar command succeded, but that is not truth:
```
$ cat /tmp/archive 
tar: Refusing to write archive contents to terminal (missing -f option?)
tar: Error is not recoverable: exiting now
```
You can observe two issues with docker here:
- First is that `tar` always puts the error output into `stderr`, so to redirect the error output you must always use `2>` instead of `>`.  IOW, even if you use '>' here, the error should be shown.  IOW(2), stdout is merged with stderr which would break a lot of things if this has been done in usual terminal.
- The second is that in fact `>` naturally causes that file descriptor `stdout` (number `1`) is no longer a terminal for the command, but regular file.  In docker however, `tar` is confused because he's still having attached (some fake) `pty` (terminal) and thus it refuses to write there.

**Comment by praiskup** [comment]
```
[root@fd09f345bb34 /]# cat /tmp/script 
#! /bin/sh

test -t 1 && green="\e[0;32m" reset="\e[0m"
echo -e "${green}Some (potentially) green output$reset"
```
Then:
```
$ docker exec -ti fd09f345bb34 /bin/bash /tmp/script 
Some (potentially) green output
```
^^ still good, the output is green, but:

```
$ docker exec -ti fd09f345bb34 /bin/bash /tmp/script > output
$ vim output
^[[0;32mSome (potentially) green output^[[0m
```
You should see only `Some (potentially) green output` there, because script should be able to detect that colorizing is not possible.

**Comment by praiskup** [comment]
@LK4D4 wrote:
>  I encourage you to open new issue and describe real use cases there.
If you found this not-a-bug for you, I can disagree and shrug, but it is probably OK -- another issue would be useless. I just wanted to help.

**Comment by justincormack** [comment]
Unfortunately while there are some simple cases where having Docker auto
detect whether to create a tty based on the command, there are more complex
use cases where it is really hard, as you can attach to a command later for
example. Given that the default is no tty, ignoring that the user specified
a tty is also a problem. You can wrap the docker usage in a script that
checks is it is on a tty fairly easily to work around this.

**Comment by praiskup** [comment]
Maybe we could have some option for that "auto" pty?  No doubts, those are valid arguments.

Can we elaborate on the 'stdout', 'stderr' merge?

**Comment by praiskup** [comment]
Yes, I'm (to some extent) aware of this, but thanks for putting it here to someone else.

If I wanted to solve this issue for my program, for something I would like to ship to customers, the shell wrapper doing `test -t` (and setting appropriate docker cli options) would work for me.   I just considered it terribly inconvenient in terminal app to do this.  It looks like the container is "too far from the actual terminal" to have this fixed easily, which is something I can accept.

**Comment by praiskup** [comment]
I.e. I can still imagine e.g. `--detect-terminal` client option "doing the right thing", is that real?

**Comment by LK4D4** [comment]
@praiskup I understand the problem now(same aliases for tty and file). I'm glad that you wrote so many comments to explain the issue. Please, include your `alias` + `coloring script` examples into issue description.
@justincormack I think with proper documentation `detect tty` feature is ok for `docker run/exec`.
